ReplicaEngine/tla/ReplicaEngine.tla

```diff
@@ -0,0 +1,832 @@
+                    deleteFromLucene := TRUE;
+                    currentlyDeleted := TRUE;
+                else
+                    if req.seqno > lucene.document.seqno
+                    then
+                        (* OP_NEWER *)
+                        plan := "processNormally";                 
+                        deleteFromLucene := TRUE;
+                        currentlyDeleted := FALSE;
+                    else
+                        (* OP_STALE_OR_EQUAL *)
+                        plan := "processButSkipLucene";                  
+                        deleteFromLucene := FALSE;
+                        currentlyDeleted := FALSE;
+                    end if;
+                end if;
+            end if;
+        end if;
+        
+        ExecuteDeletePlan:  \* Label needed because of AwaitRefreshOnDelete label
+        if deleteFromLucene
+        then
+            if currentlyDeleted = FALSE
+            then
+                lucene := [lucene EXCEPT !.buffer = Append(@, [ type |-> Lucene_deleteDocuments ])];
+            end if;
+
+            versionMap_entry := [ type |-> DELETE, seqno |-> req.seqno, flushed |-> FALSE ];
+        end if;
+    
+        completedSeqnos := completedSeqnos \cup {req.seqno};
+    else
+
+        (* planIndexingAsNonPrimary *)
+        
+        (* A RETRY_ADD has canOptimiseAddDocument = TRUE and
+            mayHaveBeenIndexedBefore = TRUE so is planned normally,
+            but also updates maxUnsafeAutoIdTimestamp within
+            mayHaveBeenIndexedBefore() *)
+
+        if req.type = RETRY_ADD
+        then
+            maxUnsafeAutoIdTimestamp := Max(maxUnsafeAutoIdTimestamp, req.autoIdTimeStamp);
+        end if;
+           
+        (* An ADD can be optimized if mayHaveBeenIndexedBefore = FALSE
+            which is calculated by comparing timestamps. *)
+        
+        if /\ req.type = ADD
+           /\ maxUnsafeAutoIdTimestamp < req.autoIdTimeStamp
+        then
+            plan                     := "optimisedAppendOnly";
+            currentNotFoundOrDeleted := TRUE;
+            useLuceneUpdateDocument  := FALSE;
+            indexIntoLucene          := TRUE;
+        else
+        
+            (* All other operations are planned normally *)
+            versionMap_needsSafeAccess := TRUE;
+            
+            if req.seqno <= localCheckPoint
+            then
+                (* OP_STALE_OR_EQUAL *)        
+                plan                     := "processButSkipLucene";
+                currentNotFoundOrDeleted := FALSE;
+                useLuceneUpdateDocument  := FALSE;
+                indexIntoLucene          := FALSE;
+            else
+                if versionMap_isUnsafe
+                then
+                    (* Perform a Lucene refresh *)
+                    AwaitRefreshOnIndex: \* Label here to allow for other concurrent activity
+                    await lucene.buffer = <<>>;
```

Comment:
same comment as above for delete, do we need to do enforceSafeAccess here?

Reply:
Yep, good catch, added.